source/extensions/filters/common/rbac/matchers.h

```diff
@@ -187,20 +187,41 @@ class PolicyMatcher : public Matcher, NonCopyable {
                 ProtobufMessage::ValidationVisitor& validation_visitor,
                 Server::Configuration::CommonFactoryContext& context)
       : permissions_(policy.permissions(), validation_visitor, context),
-        principals_(policy.principals(), context), condition_(policy.condition()) {
-    if (policy.has_condition()) {
-      expr_ = Expr::createExpression(*builder, condition_);
-    }
-  }
+        principals_(policy.principals(), context),
+        expr_wrapper_(ExprWrapper::maybeCreate(policy, builder)) {}
 
   bool matches(const Network::Connection& connection, const Envoy::Http::RequestHeaderMap& headers,
                const StreamInfo::StreamInfo&) const override;
 
 private:
   const OrMatcher permissions_;
   const OrMatcher principals_;
-  const google::api::expr::v1alpha1::Expr condition_;
-  Expr::ExpressionPtr expr_;
+
+  // A wrapper that encapsulates the CEL expression and its condition.
+  class ExprWrapper {
```

Comment:
Do you mind moving it to common::expr? This seems to be extremely error prone in other contexts since both underlying proto and sourceinfo must outlive the compiled expression. See https://github.com/envoyproxy/envoy/pull/39334